### ISSUE TITLE
Slow Speed and OutOfMemory Error

### ISSUE THREAD

**Issue by Iagmon** [derrian-distro/LoRA_Easy_Training_Scripts#242]
Hello. Can someone help me to optimize my local training speed?

I got a RTX3060 (12GB VRAM) and 32GB RAM. When I train, I get a speed of ≈ 20 to 25 seconds per item. This is kinda weird to me because when I run local Kohya I get a speed of 7 to 9 seconds per item. Similar dataset and, as far as I know, similar settings.

Also, not sure if related but if I try to train with Xformers disabled, I always get an OutOfMemory error.

Is it normal to have such an extreme difference between the two trainers?

I'm posting the .toml (converted to .txt because github doesn't support .toml) of my last attempt at training so you can see my current settings. Please let me know if there's any other info I can provide.

[TOML.txt](https://github.com/user-attachments/files/17007077/TOML.txt)

[[subsets]]
caption_extension = ".txt"
image_dir = "F:/ForgeUI/MyResources/FacialHairSDXL/dataset/20_circlegoatee"
keep_tokens = 1
name = "20"
num_repeats = 20
shuffle_caption = true

[train_mode]
train_mode = "lora"

[general_args.args]
max_data_loader_n_workers = 1
persistent_data_loader_workers = true
pretrained_model_name_or_path = "F:/ForgeUI/ForgeUI/webui/models/Stable-diffusion/sd_xl_base_1.0.safetensors"
vae = "F:/ForgeUI/ForgeUI/webui/models/VAE/sdxl_vae.safetensors"
sdxl = true
seed = 1
max_token_length = 225
prior_loss_weight = 1.0
max_train_epochs = 10
xformers = true
mixed_precision = "bf16"

[general_args.dataset_args]
resolution = [ 1024, 1024,]
batch_size = 2

[network_args.args]
network_dim = 16
network_alpha = 8.0
min_timestep = 0
max_timestep = 1000

[optimizer_args.args]
lr_scheduler_num_cycles = 1
lr_scheduler = "cosine_with_restarts"
loss_type = "l2"
learning_rate = 0.0001
unet_lr = 5e-5
max_grad_norm = 1.0
optimizer_type = "AdamW8bit"

[saving_args.args]
output_dir = "F:/ForgeUI/MyResources/FacialHairSDXL/output"
output_name = "SDXLCircleGoateeV3"
save_precision = "fp16"
save_model_as = "safetensors"

[bucket_args.dataset_args]
enable_bucket = true
min_bucket_reso = 256
max_bucket_reso = 1024
bucket_reso_steps = 64

[network_args.args.network_args]

[optimizer_args.args.optimizer_args]
weight_decay = "0.1"

**Comment by derrian-distro** [comment]
there shouldn't be a difference at all, considering they *are the same* trainer, both use the same backend, so it's likely a result of some arg being different between the two. and yes, not using xformers will drastically increase your vram usage, so ooming there is expected and normal

**Comment by Jelosus2** [comment]
Its normal that you OOM on a RTX 3060 if you don't use gradient checkpointing. Also I would recommend enabling xformers or SDPA. Also 20 repeats is not healthy, but the issue is not about that.

**Comment by Iagmon** [comment]
I understand, but can you guys give me examples of which args I should take a better look into to see if I can improve training speed?

I know this is the place to talk about the software in specific and not to ask general advice about training, but I watched a few Youtube videos on Lora training and none of them mentions args I should use to optimize training speed.

**Comment by Jelosus2** [comment]
The only optimizations to improve training speed with your VRAM for SDXL are:
- Cache Latents (negligible improvement)
- Gradient Checkpointing (biggest improvement)
- Xformers/SDPA (little but noticeable)

**Comment by Jelosus2** [comment]
Any update? I will close the issue in the next 3 days if there's no activity
